### PR TITLE
PropType fixes: Stateless component validation & softening validation to warnings

### LINKED ISSUE #124: PropTypes validation isn't occurring against Stateless Function Components
I was attempting a PR to soften the proptypes errors into warnings as we discussed on twitter but noticed this issue.  I will look into it, but thought it worth logging in the interim.

### PR DESCRIPTION
**Pull request by ctrlplusb**
Fixes #124 and #125 

I accidentally messed up my previous PR commits.  So just recreated a new cleaned up one.

### CODE REVIEW

**Comment by developit** [comment]
Ah, sorry - I commented on the other PR. Moving the comments here.

File: src/index.js [hunk 1]
```diff
@@ -129,11 +129,40 @@ for (let i=ELEMENTS.length; i--; ) {
 }
 
 
+function isStatelessComponent(c) {
```
Comment (developit) [inline src/index.js]: Thinking we should make this match [preact's existing implementation](https://github.com/developit/preact/blob/master/src/vdom/functional-component.js#L12-L15), which allows Components to not extend from Component:

```js
function isStatelessComponent(c) {
	return typeof c==='function' && !(c.prototype && c.prototype.render);
}
```

File: src/index.js [hunk 2]
```diff
@@ -129,11 +129,40 @@ for (let i=ELEMENTS.length; i--; ) {
...
+
+function statelessComponentHook(WrappedComponent) {
+	// Ensure the displayName has been set for debugging purposes
+	WrappedComponent.displayName = WrappedComponent.displayName
+		|| WrappedComponent.name
+		|| 'StatelessComponent';
+
+	function StatelessComponent(props) {
+		// Ensure default props have been applied to the props.
+		const p = WrappedComponent.defaultProps
```
Comment (developit) [inline src/index.js]: I believe defaultProps should already be applied by this point - stateless functions are invoked [here](https://github.com/developit/preact/blob/master/src/vdom/functional-component.js#L24), and their props are mutated prior to invocation [here](https://github.com/developit/preact/blob/master/src/vdom/index.js#L36-L45), which seems to apply `defaultProps`.

Thoughts?

-edit-: also, we'll want to also forward the second argument (`context`) through `StatelessComponent`, just passing it directly to `WrappedComponent`.

**Comment by developit** [comment]
Regarding the comments - I'm also fine merging and modifying after, I'll probably add a few tests for this anyway. Let me know.

**Comment by ctrlplusb** [comment]
Ah cool, you are welcome to go ahead and merge. I am just taking my daughter swimming so I'll be a while away still. :-)

**Comment by ctrlplusb** [comment]
As a note though it wasn't my experience that the default props were applied for functional comps. I tried a local install, so perhaps there is another issue somewhere? Heads up. :-)

**Comment by developit** [comment]
They weren't applied using the normal DOM renderer?

**Comment by ctrlplusb** [comment]
Nope. I was using the client side Dom render and was getting errors for props with defaults :-(

**Comment by developit** [comment]
Just published the fix for `preact-render-to-string` that adds `defaultProps`. Will look into `defaultProps` issue in core.  Since that's part of preact itself (not `preact-compat`), I want to fix it there if it's broken.

**Comment by ctrlplusb** [comment]
Sorry I should have mentioned: the issues was for stateless comps only.

**Comment by ctrlplusb** [comment]
Feel free to close this PR and work against your report directly if it's easier for you. I am not precious about having my name against the report, just wanted to help with the issues. :-)

**Comment by developit** [comment]
I'll merge, just will remove the defaultProps thing if I can figure this out haha.

**Pull request merged**